### ISSUE TITLE
Inconsistent losses for Keras

### ISSUE THREAD

**Issue by elbaro** [tensorflow/tensorflow#11178]
TF 1.2.0, Python 3.5.

## TF Official Keras
```
tf.contrib.keras.backend.sparse_categorical_crossentropy (output, target)
tf.contrib.keras.losses.sparse_categorical_crossentropy  (target, output)
tf.contrib.keras.metrics.sparse_categorical_crossentropy (target, output)
tf.losses.sparse_softmax_cross_entropy (target, output)  - no sentinel
tf.nn.sparse_softmax_cross_entropy_with_logits (target, output)
'sparse_softmax_cross_entropy'
```

## Pure Keras
```
keras.losses.sparse_categorical_crossentropy(target, output)
```

There are many `sparse_softmax_cross_entropy` but they all have different APIs. Some are compatible with `tf.contrib.keras.models.Model.compile()` and  the others are not because of `_sentinel` and different order of `output` and `target`.

Interestingly both the original TF and Keras use (target, output) and only `tf.contrib.keras.backend` version uses (output, target). tf-keras needs to use (output, target) otherwise the model is not trained.

Can we unify the interface and document when to use each version?

**Comment by ali01** [comment]
Thanks for reaching out!

cc: @martinwicke

**Comment by martinwicke** [comment]
Yes. We can remove the sentinel. We put it in as a safety since the those losses were not symmetric and we switched the argument order.

Hopefully, then, all losses will use (target, output), and can all be used with compile.

**Comment by fchollet** [comment]
> tf.contrib.keras.backend version uses (output, target)

That was modeled after the initial TF version. It was an oversight and we can change it. `backend.sparse_categorical_crossentropy` is meant as a cross-backend layer, it's not for public consumption; the official API is in `keras.losses`.

**Comment by George-Zhu** [comment]
Mark

**Comment by George-Zhu** [comment]
I also found that the input arguments order of keras.backend.categorical_crossentropy() and keras.backend.tensorflow_backend.categorical_crossentropy() are not consistent with keras.losses.categorical_crossentropy(). I found it when I try to check my own implementation of cross entropy loss function:
```
import numpy as np
import keras
import keras.backend as K

def categorical_crossentropy_np(y_true, y_pred):
    _EPSILON = K.epsilon()
    y_pred = y_pred / np.expand_dims(np.sum(y_pred, axis=-1), axis=-1)
    y_pred = np.clip(y_pred, _EPSILON, 1.0-_EPSILON)
    return -np.sum(y_true * np.log(y_pred), axis=-1)

def check_loss():
    shape = (6, 8)

    y_a = np.random.random(shape).astype(np.float32)
    y_b = np.random.random(shape).astype(np.float32)

    out1 = K.eval(keras.backend.categorical_crossentropy(K.variable(y_a), K.variable(y_b)))
    #out1 = K.eval(keras.backend.tensorflow_backend.categorical_crossentropy(K.variable(y_a), K.variable(y_b)))
    #out1 = K.eval(keras.losses.categorical_crossentropy(K.variable(y_a), K.variable(y_b)))
    out2 = categorical_crossentropy_np(y_a, y_b)

    assert out1.shape == out2.shape
    assert out1.shape == shape[:-1]

    print out1
    print out2

if __name__ == '__main__':
    check_loss()

```

**Comment by martinwicke** [comment]
The backend functions are internal only and fixable. PRs welcome.

**Comment by fchollet** [comment]
This specific issue has been fixed in Keras 2.0.7. The version included in
contrib as of TF 1.3 is Keras 2.0.6, though.

**Comment by George-Zhu** [comment]
Got it. Thanks for your reply!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by martinwicke** [comment]
We updated past 2.0.9, so this should be fixed.
